README
libspctag README file

See INSTALL file for installation instructions.

libspctag is a library written in C that can read and write ID666 tags from SPC files. For the moment, the library can not read or write extended ID666 tags.

Look at the test/test.c file to understand how this library should be used.

Author : Jérôme SONRIER <jsid@emor3j.fr.eu.org>



Changelog
== version 0.2 (2011-09-18) ==
	* Add a changelog file
	* Fix printing of text tags
	* Bump up soversion number
	* Rewrite test program
	* Add save() function
	* Move to the start of the file before working on it
	* Add functions to set length, fade length, artist, channels and emulator
	* Fix spctag_get_emulator()
	* Export spctag_is_init and spctag_txt_tag
	* Don't export spctag_is_text()
	* Fix length and fade length detection for binary tags
	* Add functions to set song title, game title, comments and dump date
	* Move tag structure declarations to spctag.h
	* Move fields length and return codes definitions to spctag.h
	* Prefix all definitions with SPCTAG 
	* Only export symbols belonging to the public API

== version 0.1 (2011-07-03) ==
	* Initial commit


test/test.c
#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <ctype.h>
#include <string.h>
#include <spctag.h>

int main( int argc, char **argv )
{
	FILE *spc_file;					// Our SPC file
	char* test_str = "abcdefghijklmnopqrstuvwxyzabcdef-azerty";
	char* test_date = "12/31/1999";
	char *test_num = "999";
	char song_title[SPCTAG_SONGTITLE_LENGTH+1];
	char game_title[SPCTAG_GAMETITLE_LENGTH+1];
	char dumper_name[SPCTAG_DUMPERNAME_LENGTH+1];
	char comments[SPCTAG_COMMENTS_LENGTH+1];
	char dump_date[SPCTAG_DUMPDATE_TXT_LENGTH+1];
	char length[SPCTAG_LENGTH_LENGTH+1];
	char fade_length[SPCTAG_FADELENGTH_BIN_LENGTH+1];
	char artist[SPCTAG_ARTIST_LENGTH+1];
	char channels[9];
	char emulator[20];

	// Exit if no filename is specify
	if ( argc < 2 ) {
		fprintf( stderr, "Usage : %s filname\n", argv[0]);
		exit( 1 );
	}

	// Open file or exit
	if ( ( spc_file = fopen( argv[1], "r+b" ) ) == NULL ) {
		fprintf( stderr, "Unable to open file '%s'!\n", argv[1] );
		exit( 2 );
	}

	spctag_init( spc_file );
	
	printf( "Version : %s\n", LIBSPCTAG_VERSION_STR );
	
	if ( spctag_txt_tag )
		printf( "Tags format: Text\n\n" );
	else
		printf( "Tags format: Binary\n\n" );

	printf( "--- Test reading tags ---\n");

	strcpy( song_title, spctag_get_songtitle() );
	strcpy( game_title, spctag_get_gametitle() );
	strcpy( dumper_name, spctag_get_dumpername() );
	strcpy( comments, spctag_get_comments() );
	strcpy( dump_date, spctag_get_dumpdate() );
	strcpy( length, spctag_get_length() );
	strcpy( fade_length, spctag_get_fadelength() );
	strcpy( artist, spctag_get_artist() );
	strcpy( channels, spctag_get_defaultchannels() );
	strcpy( emulator, spctag_get_emulator() );

	printf("Song title : %s\n", song_title );
	printf("Game title : %s\n", game_title );
	printf("Dumper name : %s\n", dumper_name );
	printf("Comments : %s\n", comments );
	printf("Dump date : %s\n", dump_date );
	printf("Length : %s\n", length );
	printf("Fade length : %s\n", fade_length );
	printf("Artist : %s\n", artist );
	printf("Default channels : %s\n", channels );
	printf("Emulator : %s\n", emulator );

	printf( "\n--- Test writing tags ---\n");

	spctag_set_songtitle( test_str );
	spctag_set_gametitle( test_str );
	spctag_set_dumpername( test_str );
	spctag_set_comments( test_str );
	spctag_set_dumpdate( test_date );
	spctag_set_length( test_num );
	spctag_set_fadelength( test_num );
	spctag_set_artist( test_str );
	spctag_set_defaultchannels( "01010101" );
	spctag_set_emulator( "Snes9x" );

	spctag_save( spc_file );

	printf("Song title : %s\n", spctag_get_songtitle() );
	printf("Game title : %s\n", spctag_get_gametitle() );
	printf("Dumper name : %s\n", spctag_get_dumpername() );
	printf("Comments : %s\n", spctag_get_comments() );
	printf("Dump date : %s\n", spctag_get_dumpdate() );
	printf("Length : %s\n", spctag_get_length() );
	printf("Fade length : %s\n", spctag_get_fadelength() );
	printf("Artist : %s\n", spctag_get_artist() );
	printf("Default channels : %s\n", spctag_get_defaultchannels() );
	printf("Emulator : %s\n", spctag_get_emulator() );

	printf( "\n--- Restor original tags ---\n");

	spctag_set_songtitle( song_title );
	spctag_set_gametitle( game_title );
	spctag_set_dumpername( dumper_name );
	spctag_set_comments( comments );
	spctag_set_dumpdate( dump_date );
	spctag_set_length( length );
	spctag_set_fadelength( fade_length );
	spctag_set_artist( artist );
	spctag_set_defaultchannels( channels );
	spctag_set_emulator( emulator );

	spctag_save( spc_file );

	printf("Song title : %s\n", spctag_get_songtitle() );
	printf("Game title : %s\n", spctag_get_gametitle() );
	printf("Dumper name : %s\n", spctag_get_dumpername() );
	printf("Comments : %s\n", spctag_get_comments() );
	printf("Dump date : %s\n", spctag_get_dumpdate() );
	printf("Length : %s\n", spctag_get_length() );
	printf("Fade length : %s\n", spctag_get_fadelength() );
	printf("Artist : %s\n", spctag_get_artist() );
	printf("Default channels : %s\n", spctag_get_defaultchannels() );
	printf("Emulator : %s\n", spctag_get_emulator() );

	// Close file
	fclose( spc_file );

	return 0;
}


src/spctag.c
/*
    spctag.c : spctag library
    v0.2 - 2011-09-18
    
    Copyright (C) 2011 Jérôme SONRIER <jsid@emor3j.fr.eu.org>

    This file is part of libspctag.

    libspctag is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    libspctag is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with libspctag.  If not, see <http://www.gnu.org/licenses/>.
*/


#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "spctag.h"
#include "constants.h"

// Returned variables
char song_title[SPCTAG_SONGTITLE_LENGTH+1];
char game_title[SPCTAG_GAMETITLE_LENGTH+1];
char dumper_name[SPCTAG_DUMPERNAME_LENGTH+1];
char comments[SPCTAG_COMMENTS_LENGTH+1];
char dump_date[SPCTAG_DUMPDATE_TXT_LENGTH+1];
char length[SPCTAG_LENGTH_LENGTH+1];
char fade_length[SPCTAG_FADELENGTH_TXT_LENGTH+1];
char artist[SPCTAG_ARTIST_LENGTH+1];
char channels[9];
char *emulator;

/*
 * get_formated_channels
 * Transform a byte into string
 * param:
 *     channels: channels encoded on a byte
 *     formated_channels: string composed of 0 and 1 that represent channels states
 */
void get_formated_channels( char channels, char* formated_channels )
{
	unsigned char masque = 0x1 << 7;
	int i;

	for ( i=0; i<=7; i++ ) {
		if ( channels & masque )
			formated_channels[i] = '1';
		else
			formated_channels[i] = '0';

		masque >>= 1;
	}

	formated_channels[i] = '\0';
}

/*
 * set_formated_channels
 * Transform a string composed of 0 and 1 into a byte
 * param:
 *     formated_channels: string composed of 0 and 1 that represent channels states
 *     channels: channels encoded on a byte
 */
void set_formated_channels( char* formated_channels, char* channels )
{
	unsigned char masque;
	int i;

	for ( i=0; i<=7; i++ ) {
		masque = 0x1 << 7-i;
		
		if ( formated_channels[i] == '1' )
			*channels |= masque;
		else {
			masque = ~masque;
			*channels &= masque;
		}
	}
}

/*
 * spctag_is_text
 * Read data and tells if it is only text
 * param:
 *     data: data wich must be test
 * return:
 *     0: at least one byte of data is not considered as text
 *     1: all bytes are considered as text
 */
int spctag_is_text( const char *data )
{
	int i;

	for ( i=0; i<SPCTAG_TEST_LENGTH; i++ ) {
		if ( data[i] != 0 && ( data[i] < 44 || data[i] > 58 ) )
			return 0;
	}

	return 1;
}

int spctag_init( FILE* file )
{
	char header[SPCTAG_FILE_HEADER_LENGTH+1];	// SPC file header
	char has_tag;
	char test_tag[SPCTAG_TEST_LENGTH+1];

	// Return if the file is not a SCP file
	fseek( file, 0, SEEK_SET );
	fread( &header, sizeof( header ) - 1, 1, file );
	header[SPCTAG_FILE_HEADER_LENGTH] = '\0';
	if ( strcmp( header, SPCTAG_FILE_HEADER ) != 0 ) {
		fprintf( stderr, "Not an %s\n", SPCTAG_FILE_HEADER );
		return SPCTAG_E_NOT_SPC;
	}

	// Return if the file header do not contain ID666 tags
	fseek( file, SPCTAG_HAS_TAGS, SEEK_SET );
	has_tag = fgetc( file );
	if ( has_tag != SPCTAG_TAGS_PRESENTS ) {
		printf( "No ID666 tag.\n" );
		return SPCTAG_E_NO_TAGS;
	}

	// Test if we have text or binary tags
	fseek( file, SPCTAG_TEST_START, SEEK_SET );
	fread( &test_tag, SPCTAG_TEST_LENGTH, 1, file );
	spctag_txt_tag = spctag_is_text( test_tag );

	// Init tag struct
	fseek( file, SPCTAG_TAGS_START, SEEK_SET );
	if ( spctag_txt_tag ) {
		spctag_tags_txt = (struct TAGS_TXT *)malloc( sizeof( struct TAGS_TXT ) );
		fread( spctag_tags_txt, sizeof( struct TAGS_TXT ), 1, file );
	} else {
		spctag_tags_bin = (struct TAGS_BIN *)malloc( sizeof( struct TAGS_BIN ) );
		fread( spctag_tags_bin, sizeof( struct TAGS_BIN ), 1, file );
	}

	spctag_is_init = 1;

	return SPCTAG_SUCCESS;
}

char* spctag_get_songtitle()
{
	if ( spctag_txt_tag )
		strncpy( song_title, spctag_tags_txt->song_title, SPCTAG_SONGTITLE_LENGTH );
	else
		strncpy( song_title, spctag_tags_bin->song_title, SPCTAG_SONGTITLE_LENGTH );

	song_title[SPCTAG_SONGTITLE_LENGTH] = '\0';

	return song_title;
}

char* spctag_get_gametitle()
{
	if ( spctag_txt_tag )
		strncpy( game_title, spctag_tags_txt->game_title, SPCTAG_GAMETITLE_LENGTH );
	else
		strncpy( game_title, spctag_tags_bin->game_title, SPCTAG_GAMETITLE_LENGTH );

	game_title[SPCTAG_GAMETITLE_LENGTH] = '\0';

	return game_title;
}

char* spctag_get_dumpername()
{
	if ( spctag_txt_tag )
		strncpy( dumper_name, spctag_tags_txt->dumper_name, SPCTAG_DUMPERNAME_LENGTH );
	else
		strncpy( dumper_name, spctag_tags_bin->dumper_name, SPCTAG_DUMPERNAME_LENGTH );
	
	dumper_name[SPCTAG_DUMPERNAME_LENGTH] = '\0';
	
	return dumper_name;
}

char* spctag_get_comments()
{
	if ( spctag_txt_tag )
		strncpy( comments, spctag_tags_txt->comments, SPCTAG_COMMENTS_LENGTH );
	else
		strncpy( comments, spctag_tags_bin->comments, SPCTAG_COMMENTS_LENGTH );

	comments[SPCTAG_COMMENTS_LENGTH] = '\0';

	return comments;
}

char* spctag_get_dumpdate()
{
	if ( spctag_txt_tag )
		strncpy( dump_date, spctag_tags_txt->dump_date, SPCTAG_DUMPDATE_TXT_LENGTH );
	else {
		if ( spctag_tags_bin->dump_day != 0 && spctag_tags_bin->dump_month != 0 && spctag_tags_bin->dump_year != 0 )
			sprintf( dump_date, "%.2d/%.2d/%.4u", spctag_tags_bin->dump_month, spctag_tags_bin->dump_day, *(unsigned*)spctag_tags_bin->dump_year );
		else
			dump_date[0] = '\0';
	}

	dump_date[SPCTAG_DUMPDATE_TXT_LENGTH] = '\0';

	return dump_date;
}

char* spctag_get_length()
{
	if ( spctag_txt_tag )
		strncpy( length, spctag_tags_txt->length, SPCTAG_LENGTH_LENGTH );
	else {
		unsigned bin_length = 0;
		strncpy( (char*)&bin_length, spctag_tags_bin->length, SPCTAG_LENGTH_LENGTH );
		sprintf( length, "%u", bin_length );
	}

	length[SPCTAG_LENGTH_LENGTH] = '\0';

	return length;
}

char* spctag_get_fadelength()
{
	if ( spctag_txt_tag )
		strncpy( fade_length, spctag_tags_txt->fade_length, SPCTAG_FADELENGTH_TXT_LENGTH );
	else {
		unsigned bin_fadelength = 0;
		strncpy( (char*)&bin_fadelength, spctag_tags_bin->fade_length, SPCTAG_FADELENGTH_BIN_LENGTH );
		sprintf( fade_length, "%u", bin_fadelength );
	}

	fade_length[SPCTAG_FADELENGTH_TXT_LENGTH] = '\0';

	return fade_length;
}

char* spctag_get_artist()
{
	if ( spctag_txt_tag )
		strncpy( artist, spctag_tags_txt->artist, SPCTAG_ARTIST_LENGTH );
	else
		strncpy( artist, spctag_tags_bin->artist, SPCTAG_ARTIST_LENGTH );

	artist[SPCTAG_ARTIST_LENGTH] = '\0';

	return artist;
}

char* spctag_get_defaultchannels()
{	
	if ( spctag_txt_tag )
		get_formated_channels( spctag_tags_txt->default_channels, channels );
	else
		get_formated_channels( spctag_tags_bin->default_channels, channels );

	return channels;
}

char* spctag_get_emulator()
{
	if ( spctag_txt_tag ) {
		switch ( spctag_tags_txt->emulator ) {
			case '1' : emulator = "ZSNES"; break;
			case '2' : emulator = "Snes9x"; break;
			default : emulator = "Unknown";
		}
	} else {
		switch ( (int)spctag_tags_bin->emulator ) {
			case 1 : emulator = "ZSNES"; break;
			case 2 : emulator = "Snes9x"; break;
			case 3 : emulator = "ZST2SPC"; break;
			case 4 : emulator = "Other"; break;
			case 5 : emulator = "SNEShout"; break;
			case 6 : emulator = "ZSNES / W"; break;
			case 7 : emulator = "Snes9xpp"; break;
			case 8 : emulator = "SNESGT"; break;
			default : emulator = "Unknown";
		}
	}

	return emulator;
}

int spctag_set_songtitle( char* new_songtitle )
{
	if ( spctag_txt_tag )
		strncpy( spctag_tags_txt->song_title, new_songtitle, SPCTAG_SONGTITLE_LENGTH );
	else
		strncpy( spctag_tags_bin->song_title, new_songtitle, SPCTAG_SONGTITLE_LENGTH );

	return SPCTAG_SUCCESS;
}

int spctag_set_gametitle( char* new_gametitle )
{
	if ( spctag_txt_tag )
		strncpy( spctag_tags_txt->game_title, new_gametitle, SPCTAG_GAMETITLE_LENGTH );
	else
		strncpy( spctag_tags_bin->game_title, new_gametitle, SPCTAG_GAMETITLE_LENGTH );

	return SPCTAG_SUCCESS;
}

int spctag_set_dumpername( char* new_dumpername )
{
	if ( spctag_txt_tag )
		strncpy( spctag_tags_txt->dumper_name, new_dumpername, SPCTAG_DUMPERNAME_LENGTH );
	else
		strncpy( spctag_tags_bin->dumper_name, new_dumpername, SPCTAG_DUMPERNAME_LENGTH );

	return SPCTAG_SUCCESS;
}

int spctag_set_comments( char* new_comments )
{
	if ( spctag_txt_tag )
		strncpy( spctag_tags_txt->comments, new_comments, SPCTAG_COMMENTS_LENGTH );
	else
		strncpy( spctag_tags_bin->comments, new_comments, SPCTAG_COMMENTS_LENGTH );

	return SPCTAG_SUCCESS;
}

int spctag_set_dumpdate( char* new_dumdate )
{
	if ( spctag_txt_tag )
		strncpy( spctag_tags_txt->dump_date, new_dumdate, SPCTAG_DUMPDATE_TXT_LENGTH );
	else {
		char day_string[3];
		char month_string[3];
		char year_string[5];
		char day, month;
		unsigned year;

		strncpy( day_string, new_dumdate+3, 2 );
		strncpy( month_string, new_dumdate, 2 );
		strncpy( year_string, new_dumdate+6, 4 );

		day = atoi( day_string ) & 0xFF;
		month = atoi( month_string ) & 0xFF;
		year = atoi( year_string );

		spctag_tags_bin->dump_day = day;
		spctag_tags_bin->dump_month = month;
		strncpy( spctag_tags_bin->dump_year, (char*)&year, SPCTAG_DUMPYEAR_BIN_LENGTH );
	}

	return SPCTAG_SUCCESS;
}

int spctag_set_length( char* new_length )
{
	if ( spctag_txt_tag )
		strncpy( spctag_tags_txt->length, new_length, SPCTAG_LENGTH_LENGTH );
	else {
		unsigned length = (unsigned)atoi( new_length );

		if ( length > 0xFFFFFF )
			length = 0xFFFFFF;

		strncpy( spctag_tags_bin->length, (char*)&length, SPCTAG_LENGTH_LENGTH );
	}

	return SPCTAG_SUCCESS;
}

int spctag_set_fadelength( char* new_fadelength )
{
	if ( spctag_txt_tag )
		strncpy( spctag_tags_txt->fade_length, new_fadelength, SPCTAG_FADELENGTH_TXT_LENGTH );
	else {
		unsigned fadelength = (unsigned)atoi( new_fadelength );

		if ( fadelength > 0xFFFFFFFF )
			fadelength = 0xFFFFFFFF;

		strncpy( spctag_tags_bin->fade_length, (char*)&fadelength, SPCTAG_FADELENGTH_BIN_LENGTH);
	}

	return SPCTAG_SUCCESS;
}

int spctag_set_artist( char* new_artist )
{
	if ( spctag_txt_tag )
		strncpy( spctag_tags_txt->artist, new_artist, SPCTAG_ARTIST_LENGTH );
	else
		strncpy( spctag_tags_bin->artist, new_artist, SPCTAG_ARTIST_LENGTH );

	return SPCTAG_SUCCESS;
}

int spctag_set_defaultchannels( char* new_defaultchannels )
{
	if ( spctag_txt_tag ) {
		set_formated_channels( new_defaultchannels, &spctag_tags_txt->default_channels );
	} else {
		set_formated_channels( new_defaultchannels, &spctag_tags_bin->default_channels );
	}
}

int spctag_set_emulator( char* new_emulator )
{
	if ( spctag_txt_tag ) {
		if ( strcasecmp( new_emulator, "ZSNES") == 0 )
			spctag_tags_txt->emulator = '1';
		else if ( strcasecmp( new_emulator, "Snes9x") == 0 )
			spctag_tags_txt->emulator = '2';
		else
			spctag_tags_txt->emulator = '0';
	} else {
		if ( strcasecmp( new_emulator, "ZSNES") == 0 )
			spctag_tags_bin->emulator = 1;
		else if ( strcasecmp( new_emulator, "Snes9x") == 0 )
			spctag_tags_bin->emulator = 2;
		else
			spctag_tags_bin->emulator = 0;
	}
}

int spctag_save( FILE* file )
{
	// Move to the start of ID666 tags
	fseek( file, SPCTAG_TAGS_START, SEEK_SET );

	// Write structures to file
	if ( spctag_txt_tag )
		fwrite( spctag_tags_txt, sizeof(*spctag_tags_txt), 1, file );
	else
		fwrite( spctag_tags_bin, sizeof(*spctag_tags_bin), 1, file );
}

void spctag_free()
{
	if ( spctag_txt_tag )
		free( spctag_tags_txt );
	else
		free( spctag_tags_bin );
}


src/constants.h
/*
    constants.h : spctag library constants definition
    v0.2 - 2011-09-18
    
    Copyright (C) 2011 Jérôme SONRIER <jsid@emor3j.fr.eu.org>

    This file is part of libspctag.

    libspctag is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    libspctag is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with libspctag.  If not, see <http://www.gnu.org/licenses/>.
*/


// SCP file header
#define SPCTAG_FILE_HEADER "SNES-SPC700 Sound File Data"
#define SPCTAG_FILE_HEADER_LENGTH 27

// Tags starts at offset 46
#define SPCTAG_TAGS_START 46

// Tags are in text format if the 19 byte at offset 158 are all only text characteres
#define SPCTAG_TEST_START 158 
#define SPCTAG_TEST_LENGTH 19

// File contains ID666 information if byte at offset 35 = 26
#define SPCTAG_HAS_TAGS 35
#define SPCTAG_TAGS_PRESENTS 26


src/spctag.h
/*
    spctag.h : interface of the spctag library
    v0.2 - 2011-09-18
    
    Copyright (C) 2011 Jérôme SONRIER <jsid@emor3j.fr.eu.org>

    This file is part of libspctag.

    libspctag is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    libspctag is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with libspctag.  If not, see <http://www.gnu.org/licenses/>.
*/


#ifndef SPCTAG_H
#define SPCTAG_H 1

/* Library version */
#define LIBSPCTAG_MAJOR		0 /* Major number of library version */
#define LIBSPCTAG_MINOR		2 /* Minor number of library version */

/* Library version (string) */
#define LIBSPCTAG_VERSION_STR	"0.2"

// Field length
#define SPCTAG_SONGTITLE_LENGTH 32
#define SPCTAG_GAMETITLE_LENGTH 32
#define SPCTAG_DUMPERNAME_LENGTH 16
#define SPCTAG_COMMENTS_LENGTH 32
#define SPCTAG_DUMPDATE_TXT_LENGTH 11
#define SPCTAG_DUMPYEAR_BIN_LENGTH 2
#define SPCTAG_LENGTH_LENGTH 3
#define SPCTAG_FADELENGTH_TXT_LENGTH 5
#define SPCTAG_FADELENGTH_BIN_LENGTH 4
#define SPCTAG_ARTIST_LENGTH 32

// Return codes
#define SPCTAG_SUCCESS 0
#define SPCTAG_E_NOT_SPC -1
#define SPCTAG_E_NO_TAGS -2

#define SPCTAG_EXPORT __attribute__ ((visibility ("default")))

/* Represents ASCII tags */
struct TAGS_TXT {
	char song_title[SPCTAG_SONGTITLE_LENGTH];		// Song title
	char game_title[SPCTAG_GAMETITLE_LENGTH];		// Game title
	char dumper_name[SPCTAG_DUMPERNAME_LENGTH];		// Name of dumper
	char comments[SPCTAG_COMMENTS_LENGTH];			// Comments
	char dump_date[SPCTAG_DUMPDATE_TXT_LENGTH];		// Date SPC was dumped (MM/DD/YYYY)
	char length[SPCTAG_LENGTH_LENGTH];			// Number of seconds to play song before fading out
	char fade_length[SPCTAG_FADELENGTH_TXT_LENGTH];		// Length of fade in milliseconds
	char artist[SPCTAG_ARTIST_LENGTH];			// Artist of song
	char default_channels;					// Default channel disables (0 = enable, 1 = disable)
	char emulator;						// Emulator used to dump SPC
	char reserved[45];
} __attribute__((packed)) *spctag_tags_txt;

/* Represents binary tags */
struct TAGS_BIN {
	char song_title[SPCTAG_SONGTITLE_LENGTH];		// Song title
	char game_title[SPCTAG_GAMETITLE_LENGTH];		// Game title
	char dumper_name[SPCTAG_DUMPERNAME_LENGTH];		// Name of dumper
	char comments[SPCTAG_COMMENTS_LENGTH];			// Comments
	char dump_day;						// Day SPC was dumped
	char dump_month;					// Month SPC was dumped
	char dump_year[SPCTAG_DUMPYEAR_BIN_LENGTH];		// Year SPC was dumped
	char unused[7];
	char length[SPCTAG_LENGTH_LENGTH];			// Number of seconds to play song before fading out
	char fade_length[SPCTAG_FADELENGTH_BIN_LENGTH];	// Length of fade in milliseconds
	char artist[SPCTAG_ARTIST_LENGTH];			// Artist of song
	char default_channels;					// Default channel disables (0 = enable, 1 = disable)
	char emulator;						// Emulator used to dump SPC
	char reserved[46];
} __attribute__((packed)) *spctag_tags_bin;

// Globale variables
SPCTAG_EXPORT int spctag_is_init = 0;		// 1 when spctag is initialized and can be used
SPCTAG_EXPORT int spctag_txt_tag = 0;		// 1 if tags are in ASCII format, 0 if tags are in binary format

/*
 * spctag_init
 * Read tags from file.
 * param:
 *     file: SPC file descriptor
 * return:
 *     0: succes
 *     <0: error
 */
SPCTAG_EXPORT int spctag_init( FILE *file );

/*
 * spctag_get_songtitle
 * Return the song title.
 * return:
 *     song title
 */
SPCTAG_EXPORT char* spctag_get_songtitle();

/*
 * spctag_get_gametitle
 * Return the game title.
 * return:
 *     game title
 */
SPCTAG_EXPORT char* spctag_get_gametitle();

/*
 * spctag_get_dumpername
 * Return the dumper name.
 * return:
 *     dumper name
 */
SPCTAG_EXPORT char* spctag_get_dumpername();

/*
 * spctag_get_comments
 * Return comments.
 * return:
 *     comments
 */
SPCTAG_EXPORT char* spctag_get_comments();

/*
 * spctag_get_dumpdate
 * Return the dump date.
 * return:
 * dump date
 */
SPCTAG_EXPORT char* spctag_get_dumpdate();

/*
 * spctag_get_length
 * Return the number of seconds to play song before fading out.
 * return:
 *     song length
 */
SPCTAG_EXPORT char* spctag_get_length();

/*
 * spctag_get_fadelength
 * Return the length of fade in milliseconds.
 * return:
 *     fade length 
 */
SPCTAG_EXPORT char* spctag_get_fadelength();

/*
 * spctag_get_artist
 * Return the artist name.
 * return:
 *     artist name
 */
SPCTAG_EXPORT char* spctag_get_artist();

/*
 * spctag_get_defaultchannels
 * Return channel states.
 * return:
 *     channel states
 */
SPCTAG_EXPORT char* spctag_get_defaultchannels();

/*
 * spctag_get_emulator
 * Return the name of the emulator used to dump.
 * return:
 *     emulator
 */
SPCTAG_EXPORT char* spctag_get_emulator();

/*
 * spctag_set_songtitle
 * Set the song title into tags structure.
 * param:
 *     new_songtitle: new song title
 * return:
 *     0: succes
 *     <0: error
 */
SPCTAG_EXPORT int spctag_set_songtitle( char* new_songtitle );

/*
 * spctag_set_gametitle
 * Set the game title.
 * param:
 *     new_gametitle: new game title
 * return:
 *     0: succes
 *     <0: error
 */
SPCTAG_EXPORT int spctag_set_gametitle( char* new_gametitle );

/*
 * spctag_set_dumpername
 * Set the dumper name.
 * param:
 *     new_dumpername: new dumper name
 * return:
 *     0: succes
 *     <0: error
 */
SPCTAG_EXPORT int spctag_set_dumpername( char* new_dumpername );

/*
 * spctag_set_comments
 * Set comments.
 * param:
 *     new_comments: new comments
 * return:
 *     0: succes
 *     <0: error
 */
SPCTAG_EXPORT int spctag_set_comments( char* new_comments );

/*
 * spctag_set_dumpdate
 * Set the dump date.
 * param:
 *     new_comments: new comments
 * return:
 *     0: succes
 *     <0: error
 */
SPCTAG_EXPORT int spctag_set_dumpdate( char* new_dumdate );

/*
 * spctag_set_length
 * Set the number of seconds to play song before fading out.
 * param:
 *     new_length: new song length
 * return:
 *     0: succes
 *     <0: error
 */
SPCTAG_EXPORT int spctag_set_length( char* new_length );

/*
 * spctag_set_fadelength
 * Set the length of fade in milliseconds.
 * param:
 *     new_fadelength: new fade length
 * return:
 *     0: succes
 *     <0: error
 */
SPCTAG_EXPORT int spctag_set_fadelength( char* new_fadelength );

/*
 * spctag_set_artist
 * Set the artist name.
 * param:
 *     new_artist: new artist name
 * return:
 *     0: succes
 *     <0: error
 */
SPCTAG_EXPORT int spctag_set_artist( char* new_artist );

/*
 * spctag_set_defaultchannels
 * Set channels states.
 * param:
 *     new_defaultchannels: new default channel states
 * return:
 *     0: succes
 *     <0: error
 */
SPCTAG_EXPORT int spctag_set_defaultchannels( char* new_defaultchannels );

/*
 * spctag_set_emulator
 * Set the name of the emulator used to dump.
 * param:
 *     new_emulator: new emulator
 * return:
 *     0: succes
 *     <0: error
 */
SPCTAG_EXPORT int spctag_set_emulator( char* new_emulator );

/*
 * spctag_save
 * Write structures into file.
 * return:
 *     0: succes
 *     <0: error
 */
SPCTAG_EXPORT int spctag_save( FILE* file );

/*
 * spctag_free
 * Free memory used by libspctag.
 */
SPCTAG_EXPORT void spctag_free();

#endif
